src/seat/keyboard/repeat.rs

```diff
@@ -17,32 +15,28 @@ use super::{
 };
 use crate::seat::SeatState;
 
-/// Internal repeat message sent to the repeating mechanism.
-#[derive(Debug)]
-pub(crate) enum RepeatMessage {
-    /// Stop the key repeat.
-    StopRepeat,
-
-    /// The key event should not have any time added, the repeating mechanism is responsible
-    /// for that instead.
-    StartRepeat(KeyEvent),
+pub(crate) struct RepeatedKey {
+    pub(crate) key: KeyEvent,
+    /// Whether this is the first event of the repeat sequence.
+    pub(crate) is_first: bool,
+}
 
-    /// Key has changed during the repeat, but the repeat shouldn't stop.
-    KeyChanged(KeyEvent),
+pub type RepeatCallback<T> = Box<dyn FnMut(&mut T, &WlKeyboard, KeyEvent) + 'static>;
 
-    /// The repeat info has changed.
-    RepeatInfo(RepeatInfo),
+pub(crate) struct RepeatData<T> {
+    pub(crate) current_repeat: Option<RepeatedKey>,
+    pub(crate) repeat_info: RepeatInfo,
+    pub(crate) loop_handle: LoopHandle<'static, T>,
```

Comment:
I think this will fail to compile if the `calloop` feature is disabled.

Reply:
ya, but the repeat module is feature gated. I did check just now though and it seems another feature gate is missing.